core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java

```diff
@@ -83,33 +107,40 @@ public GrpcManagedChannelPool() {
   }
 
   private synchronized void destroyChannels() {
+    int channelCount = 0;
+    int destroyedCount = 0;
+    List<Pair<ChannelKey, ManagedChannelReference>> channelsToDestroy = new ArrayList<>();
     try (LockResource lockExclusive = new LockResource(mLock.writeLock())) {
-      int iDestroyed = 0;
+      channelCount = mChannels.size();
       for (HashMap.Entry<ChannelKey, ManagedChannelReference> channelEntry : mChannels.entrySet()) {
         ChannelKey channelKey = channelEntry.getKey();
         ManagedChannelReference channelReference = channelEntry.getValue();
         if (channelReference.getRefCount() <= 0) {
-          ManagedChannel channel = channelReference.get();
-          channel.shutdown();
-          try {
-            channel.awaitTermination(
-                Configuration.getMs(PropertyKey.MASTER_GRPC_CHANNEL_SHUTDOWN_TIMEOUT),
-                TimeUnit.MILLISECONDS);
-          } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            // Allow thread to exit.
-          } finally {
-            channel.shutdownNow();
-          }
-          Verify.verify(channel.isShutdown());
           mChannels.remove(channelKey);
-          iDestroyed++;
-          LOG.debug("Destroyed gRPC managed channel for key:{}", channelKey);
+          channelsToDestroy.add(new Pair(channelKey, channelReference));
         }
       }
-      LOG.debug("gRPC channel terminator shut down {} channels. Remaining active channel count: {}",
-          iDestroyed, mChannels.size());
     }
+
+    for (Pair<ChannelKey, ManagedChannelReference> channelPair : channelsToDestroy) {
+      ManagedChannel channel = channelPair.getSecond().get();
+      channel.shutdown();
+      try {
```

Comment:
Consider calling shutdown on all the channels first then calling await on all of them. That should make the execution parallel.

Reply:
One of the reasons behind pooling is to serialize channel destruction as I was issues when channels are destructed concurrently. I'd like to postpone this as later improvement task. It's not critical as we don't expect many unique channels per process.

Reply:
Let's leave a TODO then